src/ray/raylet/scheduling_queue.h

```diff
@@ -37,6 +37,118 @@ enum class TaskState {
   INFEASIBLE
 };
 
+class TaskQueue {
+ public:
+  /// Creating a task queue.
+  TaskQueue() {}
+
+  /// Destructor for task queue.
+  ~TaskQueue();
+
+  /// \brief Append a task to queue.
+  ///
+  /// \param task_id The task ID for the task to append.
+  /// \param task The task to append to the queue.
+  /// \return Whether the append operation succeeds.
+  bool AppendTask(const TaskID &task_id, const Task &task);
+
+  /// \brief Remove a task from queue.
+  ///
+  /// \param task_id The task ID for the task to remove from the queue.
+  /// \param removed_tasks If the task specified by task_id is successfully
+  ///  removed from the queue, the task data is appended to the vector. Can
+  ///  be a nullptr, in which case nothing is appended.
+  /// \return Whether the removal succeeds.
+  bool RemoveTask(const TaskID &task_id,
+                  std::vector<Task> *removed_tasks = nullptr);
+
+  /// \brief Check if the queue contains a specific task id.
+  ///
+  /// \param task_id The task ID for the task.
+  /// \return Whether the task_id exists in this queue.
+  bool HasTask(const TaskID &task_id) const;
+
+  /// \brief Remove the task list of the queue.
+  /// \return A list of tasks contained in this queue.
+  const std::list<Task> &GetTasks() const;
+
+  /// \brief Recompute the resources of the tasks with minimum requirements
+  /// in the ready queue.
+  void RecomputeMinResources();
+
+protected:
+ /// A list of tasks.
+ std::list<Task> task_list_;
+ /// A hash to speed up looking up a task.
+ std::unordered_map<TaskID, std::list<Task>::iterator> task_map_;
+};
+
+class ReadyQueue : public TaskQueue {
+public:
+  /// \brief Append a task to the ready queue.
+  ///
+  /// \param task_id The task ID for the task to append.
+  /// \param task The task to append to the queue.
+  /// \return Whether the append operation succeeds.
+  bool AppendTask(const TaskID &task_id, const Task &task);
+
+  /// \brief Remove a task from the ready queue.
+  ///
+  /// \param task_id The task ID for the task to remove from the queue.
+  /// \param removed_tasks If the task specified by task_id is successfully
+  ///  removed from the queue, the task data is appended to the vector. Can
+  ///  be a nullptr, in which case nothing is appended.
+  /// \return Whether the removal succeeds.
+  bool RemoveTask(const TaskID &task_id,
+                  std::vector<Task> *removed_tasks = nullptr);
+
+  /// \brief Check whether any of the tasks in the ready queue can be scheduled.
+  ///
+  /// \param local_available_resources: resources available on local node.
+  /// \return True if there is at least a task that can be scheduled, i.e.
+  /// whose requirements can be satisfied by the local node; false otherwise.
+  const bool CanScheduleMinTask(const ResourceIdSet &local_available_resources) const;
```

Comment:
This method seems unnecessary since you can always do this from the caller by calling `GetMinTaskResources()`

Reply:
in the current implementation, true. But as this needs to handle the multi-dimensional (`k-d`) resource space, I actually think we will need a function like this. The reason for this is that a single `ResourceSet min_task_resources_` will not be enough. There's a bit more bookkeeping that should happen, because we're effectively doing an "orthogonal range query" here.